File: docs/beginner/module-04-simple-queries/code-samples/javascript/01_basic_queries.js
#!/usr/bin/env node

/**
 * Module 4: Simple Queries and Filters - Basic Queries
 * ====================================================
 * 
 * This script demonstrates basic text search operations in Azure AI Search using JavaScript.
 * Learn how to perform simple searches, use query operators, and work with search fields.
 * 
 * Prerequisites:
 * - Azure AI Search service configured
 * - Sample index with data (from previous modules)
 * - Environment variables set in .env file
 * 
 * Author: Azure AI Search Tutorial
 */

import { SearchClient, AzureKeyCredential } from '@azure/search-documents';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

/**
 * Create and return an Azure AI Search client.
 * @returns {SearchClient} Configured search client
 */
function createSearchClient() {
    const endpoint = process.env.AZURE_SEARCH_SERVICE_ENDPOINT;
    const apiKey = process.env.AZURE_SEARCH_API_KEY;
    const indexName = process.env.AZURE_SEARCH_INDEX_NAME;

    if (!endpoint || !apiKey || !indexName) {
        throw new Error('Missing required environment variables. Check your .env file.');
    }

    return new SearchClient(endpoint, indexName, new AzureKeyCredential(apiKey));
}

/**
 * Display search results in a formatted way.
 * @param {Array} results - Array of search result documents
 * @param {string} title - Title for the result set
 * @param {number} maxResults - Maximum number of results to display
 */
function displayResults(results, title, maxResults = 5) {
    console.log(`\n${'='.repeat(60)}`);
    console.log(title);
    console.log('='.repeat(60));

    if (!results || results.length === 0) {
        console.log('No results found.');
        return;
    }

    const displayCount = Math.min(results.length, maxResults);
    
    for (let i = 0; i < displayCount; i++) {
        const result = results[i];
        const document = result.document;
        
        console.log(`\n${i + 1}. ${document.title || 'No title'}`);
        console.log(`   Score: ${result.score?.toFixed(2) || 'N/A'}`);
        console.log(`   ID: ${document.id || 'N/A'}`);
        
        // Show content preview if available
        if (document.content) {
            const preview = document.content.length > 100 
                ? document.content.substring(0, 100) + '...' 
                : document.content;
            console.log(`   Preview: ${preview}`);
        }
    }

    if (results.length > maxResults) {
        console.log(`\n... and ${results.length - maxResults} more results`);
    }
}

/**
 * Convert search results async iterator to array.
 * @param {SearchResults} searchResults - Search results from Azure AI Search
 * @returns {Promise<Array>} Array of search results
 */
async function resultsToArray(searchResults) {
    const results = [];
    for await (const result of searchResults.results) {
        results.push(result);
    }
    return results;
}

/**
 * Demonstrate basic text search operations.
 * @param {SearchClient} searchClient - Azure AI Search client
 */
async function basicTextSearch(searchClient) {
    console.log('\n' + '='.repeat(80));
    console.log('BASIC TEXT SEARCH EXAMPLES');
    console.log('='.repeat(80));

    // Example 1: Simple text search
    console.log('\n1. Simple Text Search');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure');
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: 'azure'");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
        return;
    }

    // Example 2: Multi-word search
    console.log('\n2. Multi-word Search');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('machine learning');
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: 'machine learning'");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 3: Empty search (returns all documents)
    console.log('\n3. Empty Search (All Documents)');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('*', { top: 3 });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: '*' (all documents)", 3);
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }
}

/**
 * Demonstrate searching in specific fields.
 * @param {SearchClient} searchClient - Azure AI Search client
 */
async function searchWithFields(searchClient) {
    console.log('\n' + '='.repeat(80));
    console.log('FIELD-SPECIFIC SEARCH EXAMPLES');
    console.log('='.repeat(80));

    // Example 1: Search in title field only
    console.log('\n1. Search in Title Field Only');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('python', {
            searchFields: ['title']
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: 'python' in title field");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 2: Search in multiple specific fields
    console.log('\n2. Search in Multiple Fields');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure', {
            searchFields: ['title', 'content']
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: 'azure' in title and content fields");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 3: Search with field boosting (title field weighted more)
    console.log('\n3. Search with Field Boosting');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('tutorial', {
            searchFields: ['title^3', 'content'] // Title matches weighted 3x
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: 'tutorial' with title boosting (3x)");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }
}

/**
 * Demonstrate query operators in simple syntax.
 * @param {SearchClient} searchClient - Azure AI Search client
 */
async function queryOperators(searchClient) {
    console.log('\n' + '='.repeat(80));
    console.log('QUERY OPERATORS EXAMPLES');
    console.log('='.repeat(80));

    // Example 1: Required term (+)
    console.log('\n1. Required Term (+)');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('+azure search');
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: '+azure search' (azure is required)");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 2: Excluded term (-)
    console.log('\n2. Excluded Term (-)');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure -cognitive');
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: 'azure -cognitive' (exclude cognitive)");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 3: Exact phrase ("")
    console.log('\n3. Exact Phrase Search');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('"machine learning"');
        const results = await resultsToArray(searchResults);
        displayResults(results, 'Search: "machine learning" (exact phrase)');
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 4: Wildcard search (*)
    console.log('\n4. Wildcard Search');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('develop*');
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: 'develop*' (wildcard)");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 5: Grouping with parentheses
    console.log('\n5. Grouping with Parentheses');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('(azure OR microsoft) search');
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search: '(azure OR microsoft) search'");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }
}

/**
 * Demonstrate different search modes and query types.
 * @param {SearchClient} searchClient - Azure AI Search client
 */
async function searchModesAndTypes(searchClient) {
    console.log('\n' + '='.repeat(80));
    console.log('SEARCH MODES AND QUERY TYPES');
    console.log('='.repeat(80));

    // Example 1: Any search mode (default)
    console.log('\n1. Search Mode: Any (Default)');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure machine learning', {
            searchMode: 'any'
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search mode 'any': matches any term");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 2: All search mode
    console.log('\n2. Search Mode: All');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure machine learning', {
            searchMode: 'all'
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Search mode 'all': matches all terms");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 3: Simple query type (default)
    console.log('\n3. Query Type: Simple (Default)');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure AND search', {
            queryType: 'simple'
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Simple query type with AND operator");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 4: Full Lucene query type
    console.log('\n4. Query Type: Full Lucene');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('title:azure AND content:search', {
            queryType: 'full'
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Full Lucene query with field-specific search");
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }
}

/**
 * Demonstrate advanced text search features.
 * @param {SearchClient} searchClient - Azure AI Search client
 */
async function advancedTextFeatures(searchClient) {
    console.log('\n' + '='.repeat(80));
    console.log('ADVANCED TEXT SEARCH FEATURES');
    console.log('='.repeat(80));

    // Example 1: Case sensitivity (searches are case-insensitive by default)
    console.log('\n1. Case Insensitive Search');
    console.log('-'.repeat(40));

    try {
        const resultsLower = await resultsToArray(await searchClient.search('azure'));
        const resultsUpper = await resultsToArray(await searchClient.search('AZURE'));

        console.log(`Search 'azure': ${resultsLower.length} results`);
        console.log(`Search 'AZURE': ${resultsUpper.length} results`);
        console.log('Note: Both searches return the same results (case-insensitive)');
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 2: Search with top parameter
    console.log('\n2. Search with Top Parameter');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure machine learning', {
            top: 3 // Limit to top 3 results
        });
        const results = await resultsToArray(searchResults);
        displayResults(results, "Top 3 results for 'azure machine learning'", 3);
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }

    // Example 3: Include total count
    console.log('\n3. Include Total Count');
    console.log('-'.repeat(40));

    try {
        const searchResults = await searchClient.search('azure', {
            includeTotalCount: true,
            top: 5
        });

        const results = await resultsToArray(searchResults);
        const totalCount = searchResults.count;

        console.log(`Total matching documents: ${totalCount}`);
        console.log(`Returned documents: ${results.length}`);
        displayResults(results, "Sample results with total count", 3);
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }
}

/**
 * Demonstrate working with search result metadata.
 * @param {SearchClient} searchClient - Azure AI Search client
 */
async function demonstrateResultMetadata(searchClient) {
    console.log('\n' + '='.repeat(80));
    console.log('SEARCH RESULT METADATA');
    console.log('='.repeat(80));

    try {
        const searchResults = await searchClient.search('azure', { top: 3 });
        const results = await resultsToArray(searchResults);

        console.log('\nDetailed Result Analysis:');
        console.log('-'.repeat(40));

        for (let i = 0; i < results.length; i++) {
            const result = results[i];
            const document = result.document;

            console.log(`\nResult ${i + 1}:`);
            console.log(`  Document ID: ${document.id || 'N/A'}`);
            console.log(`  Search Score: ${result.score?.toFixed(4) || 'N/A'}`);
            console.log(`  Title: ${document.title || 'N/A'}`);

            // Show all available fields
            console.log('  Available fields:');
            for (const [key, value] of Object.entries(document)) {
                if (!key.startsWith('@search')) {
                    const fieldValue = value?.toString() || 'null';
                    const fieldPreview = fieldValue.length > 50 
                        ? fieldValue.substring(0, 50) + '...' 
                        : fieldValue;
                    console.log(`    ${key}: ${fieldPreview}`);
                }
            }
        }
    } catch (error) {
        console.error(`Search failed: ${error.message}`);
    }
}

/**
 * Main function to run all basic query examples.
 */
async function main() {
    console.log('Azure AI Search - Basic Queries Examples');
    console.log('='.repeat(80));

    try {
        // Create search client
        const searchClient = createSearchClient();
        console.log(`✅ Connected to search service: ${process.env.AZURE_SEARCH_SERVICE_ENDPOINT}`);
        console.log(`✅ Using index: ${process.env.AZURE_SEARCH_INDEX_NAME}`);

        // Run examples
        await basicTextSearch(searchClient);
        await searchWithFields(searchClient);
        await queryOperators(searchClient);
        await searchModesAndTypes(searchClient);
        await advancedTextFeatures(searchClient);
        await demonstrateResultMetadata(searchClient);

        console.log('\n' + '='.repeat(80));
        console.log('✅ All basic query examples completed successfully!');
        console.log('='.repeat(80));

        console.log('\n📚 What you learned:');
        console.log('• How to perform simple text searches');
        console.log('• How to search in specific fields');
        console.log('• How to use query operators (+, -, "", *, ())');
        console.log('• How to work with search modes and query types');
        console.log('• How to access search result metadata');

        console.log('\n🔗 Next steps:');
        console.log('• Run 02_filtering.js to learn about OData filters');
        console.log('• Experiment with your own search terms');
        console.log('• Try different field combinations');

    } catch (error) {
        if (error.message.includes('Missing required environment variables')) {
            console.error(`❌ Configuration error: ${error.message}`);
            console.log('\n🔧 Setup required:');
            console.log('1. Create a .env file with your Azure AI Search credentials');
            console.log('2. Ensure you have completed previous modules to create sample indexes');
            process.exit(1);
        } else {
            console.error(`❌ Unexpected error: ${error.message}`);
            process.exit(1);
        }
    }
}

// Run the main function
main().catch(console.error);

File: docs/beginner/module-04-simple-queries/code-samples/python/01_basic_queries.py
#!/usr/bin/env python3
"""
Module 4: Simple Queries and Filters - Basic Queries
====================================================

This script demonstrates basic text search operations in Azure AI Search using Python.
Learn how to perform simple searches, use query operators, and work with search fields.

Prerequisites:
- Azure AI Search service configured
- Sample index with data (from previous modules)
- Environment variables set in .env file

Author: Azure AI Search Tutorial
"""

import os
import sys
from typing import List, Dict, Any, Optional
from azure.search.documents import SearchClient
from azure.core.credentials import AzureKeyCredential
from azure.core.exceptions import HttpResponseError
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def create_search_client() -> SearchClient:
    """
    Create and return an Azure AI Search client.
    
    Returns:
        SearchClient: Configured search client
    """
    endpoint = os.getenv("AZURE_SEARCH_SERVICE_ENDPOINT")
    api_key = os.getenv("AZURE_SEARCH_API_KEY")
    index_name = os.getenv("AZURE_SEARCH_INDEX_NAME")
    
    if not all([endpoint, api_key, index_name]):
        raise ValueError("Missing required environment variables. Check your .env file.")
    
    return SearchClient(
        endpoint=endpoint,
        index_name=index_name,
        credential=AzureKeyCredential(api_key)
    )

def print_results(results: List[Dict[str, Any]], title: str, max_results: int = 5) -> None:
    """
    Print search results in a formatted way.
    
    Args:
        results: List of search result documents
        title: Title for the result set
        max_results: Maximum number of results to display
    """
    print(f"\n{'='*60}")
    print(f"{title}")
    print(f"{'='*60}")
    
    if not results:
        print("No results found.")
        return
    
    for i, result in enumerate(results[:max_results]):
        print(f"\n{i+1}. {result.get('title', 'No title')}")
        print(f"   Score: {result.get('@search.score', 'N/A'):.2f}")
        print(f"   ID: {result.get('id', 'N/A')}")
        
        # Show content preview if available
        content = result.get('content', '')
        if content:
            preview = content[:100] + "..." if len(content) > 100 else content
            print(f"   Preview: {preview}")
    
    if len(results) > max_results:
        print(f"\n... and {len(results) - max_results} more results")

def basic_text_search(search_client: SearchClient) -> None:
    """
    Demonstrate basic text search operations.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("BASIC TEXT SEARCH EXAMPLES")
    print("="*80)
    
    # Example 1: Simple text search
    print("\n1. Simple Text Search")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text="azure"))
        print_results(results, "Search: 'azure'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
        return
    
    # Example 2: Multi-word search
    print("\n2. Multi-word Search")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text="machine learning"))
        print_results(results, "Search: 'machine learning'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Empty search (returns all documents)
    print("\n3. Empty Search (All Documents)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text="*", top=3))
        print_results(results, "Search: '*' (all documents)", max_results=3)
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def search_with_fields(search_client: SearchClient) -> None:
    """
    Demonstrate searching in specific fields.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("FIELD-SPECIFIC SEARCH EXAMPLES")
    print("="*80)
    
    # Example 1: Search in title field only
    print("\n1. Search in Title Field Only")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="python",
            search_fields=["title"]
        ))
        print_results(results, "Search: 'python' in title field")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Search in multiple specific fields
    print("\n2. Search in Multiple Fields")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="azure",
            search_fields=["title", "content"]
        ))
        print_results(results, "Search: 'azure' in title and content fields")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Search with field boosting (title field weighted more)
    print("\n3. Search with Field Boosting")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="tutorial",
            search_fields=["title^3", "content"]  # Title matches weighted 3x
        ))
        print_results(results, "Search: 'tutorial' with title boosting (3x)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def query_operators(search_client: SearchClient) -> None:
    """
    Demonstrate query operators in simple syntax.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("QUERY OPERATORS EXAMPLES")
    print("="*80)
    
    # Example 1: Required term (+)
    print("\n1. Required Term (+)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text="+azure search"))
        print_results(results, "Search: '+azure search' (azure is required)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Excluded term (-)
    print("\n2. Excluded Term (-)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text="azure -cognitive"))
        print_results(results, "Search: 'azure -cognitive' (exclude cognitive)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Exact phrase ("")
    print("\n3. Exact Phrase Search")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text='"machine learning"'))
        print_results(results, 'Search: "machine learning" (exact phrase)')
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 4: Wildcard search (*)
    print("\n4. Wildcard Search")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text="develop*"))
        print_results(results, "Search: 'develop*' (wildcard)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 5: Grouping with parentheses
    print("\n5. Grouping with Parentheses")
    print("-" * 40)
    
    try:
        results = list(search_client.search(search_text="(azure OR microsoft) search"))
        print_results(results, "Search: '(azure OR microsoft) search'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def search_modes_and_types(search_client: SearchClient) -> None:
    """
    Demonstrate different search modes and query types.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("SEARCH MODES AND QUERY TYPES")
    print("="*80)
    
    # Example 1: Any search mode (default)
    print("\n1. Search Mode: Any (Default)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="azure machine learning",
            search_mode="any"
        ))
        print_results(results, "Search mode 'any': matches any term")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: All search mode
    print("\n2. Search Mode: All")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="azure machine learning",
            search_mode="all"
        ))
        print_results(results, "Search mode 'all': matches all terms")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Simple query type (default)
    print("\n3. Query Type: Simple (Default)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="azure AND search",
            query_type="simple"
        ))
        print_results(results, "Simple query type with AND operator")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 4: Full Lucene query type
    print("\n4. Query Type: Full Lucene")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="title:azure AND content:search",
            query_type="full"
        ))
        print_results(results, "Full Lucene query with field-specific search")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def advanced_text_features(search_client: SearchClient) -> None:
    """
    Demonstrate advanced text search features.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("ADVANCED TEXT SEARCH FEATURES")
    print("="*80)
    
    # Example 1: Case sensitivity (searches are case-insensitive by default)
    print("\n1. Case Insensitive Search")
    print("-" * 40)
    
    try:
        results_lower = list(search_client.search(search_text="azure"))
        results_upper = list(search_client.search(search_text="AZURE"))
        
        print(f"Search 'azure': {len(results_lower)} results")
        print(f"Search 'AZURE': {len(results_upper)} results")
        print("Note: Both searches return the same results (case-insensitive)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Minimum should match
    print("\n2. Search with Top Parameter")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="azure machine learning",
            top=3  # Limit to top 3 results
        ))
        print_results(results, "Top 3 results for 'azure machine learning'", max_results=3)
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Include total count
    print("\n3. Include Total Count")
    print("-" * 40)
    
    try:
        results = search_client.search(
            search_text="azure",
            include_total_count=True,
            top=5
        )
        
        result_list = list(results)
        total_count = results.get_count()
        
        print(f"Total matching documents: {total_count}")
        print(f"Returned documents: {len(result_list)}")
        print_results(result_list, "Sample results with total count", max_results=3)
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def demonstrate_result_metadata(search_client: SearchClient) -> None:
    """
    Demonstrate working with search result metadata.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("SEARCH RESULT METADATA")
    print("="*80)
    
    try:
        results = list(search_client.search(
            search_text="azure",
            top=3
        ))
        
        print("\nDetailed Result Analysis:")
        print("-" * 40)
        
        for i, result in enumerate(results):
            print(f"\nResult {i+1}:")
            print(f"  Document ID: {result.get('id', 'N/A')}")
            print(f"  Search Score: {result.get('@search.score', 'N/A'):.4f}")
            print(f"  Title: {result.get('title', 'N/A')}")
            
            # Show all available fields
            print("  Available fields:")
            for key, value in result.items():
                if not key.startswith('@search'):
                    field_preview = str(value)[:50] + "..." if len(str(value)) > 50 else str(value)
                    print(f"    {key}: {field_preview}")
                    
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def main():
    """
    Main function to run all basic query examples.
    """
    print("Azure AI Search - Basic Queries Examples")
    print("=" * 80)
    
    try:
        # Create search client
        search_client = create_search_client()
        print(f"✅ Connected to search service: {os.getenv('AZURE_SEARCH_SERVICE_ENDPOINT')}")
        print(f"✅ Using index: {os.getenv('AZURE_SEARCH_INDEX_NAME')}")
        
        # Run examples
        basic_text_search(search_client)
        search_with_fields(search_client)
        query_operators(search_client)
        search_modes_and_types(search_client)
        advanced_text_features(search_client)
        demonstrate_result_metadata(search_client)
        
        print("\n" + "="*80)
        print("✅ All basic query examples completed successfully!")
        print("="*80)
        
        print("\n📚 What you learned:")
        print("• How to perform simple text searches")
        print("• How to search in specific fields")
        print("• How to use query operators (+, -, \"\", *, ())")
        print("• How to work with search modes and query types")
        print("• How to access search result metadata")
        
        print("\n🔗 Next steps:")
        print("• Run 02_filtering.py to learn about OData filters")
        print("• Experiment with your own search terms")
        print("• Try different field combinations")
        
    except ValueError as e:
        print(f"❌ Configuration error: {e}")
        print("\n🔧 Setup required:")
        print("1. Create a .env file with your Azure AI Search credentials")
        print("2. Ensure you have completed previous modules to create sample indexes")
        sys.exit(1)
        
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()